File: runtime/botmanager/src/com/newbound/net/service/Service.java
package com.newbound.net.service;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.SocketAddress;
import java.net.SocketException;
import java.net.SocketTimeoutException;
import java.util.Hashtable;

import com.newbound.robot.Callback;
import com.newbound.thread.ThreadHandler;
import com.newbound.util.IsRunning;

public class Service
{
	public Container CONTAINER = null;

	protected boolean RUNNING = false;
	protected String NAME = null;

	public IsRunning ISRUNNING = new IsRunning() 
	{
		@Override
		public boolean isRunning() 
		{
			return RUNNING;
		}
	};
	
	protected Hashtable<Object, Callback> callbacks = new Hashtable();
	protected ServerSocket SS;
	
	public Service(final ServerSocket ss, final String name, final Class parser, Container th) throws IOException 
	{
		final Service me = this;
		
		SS = ss;
		CONTAINER = th;
		NAME = name;
		
		th.find("botmanager").addJob(new Runnable() 
		{
			public void run() 
			{
				RUNNING = true;
				System.out.println("Service "+name+" starting up...");
				while (RUNNING) try 
				{ 
					final Socket s = ss.accept();
					CONTAINER.getDefault().addJob(new Runnable() {
						@Override
						public void run() {
							try
							{
								Parser p = (Parser) (parser.getDeclaredConstructor().newInstance());
								if (p.init(me, s)) listen(s, p);
							}
							catch (Exception x) { x.printStackTrace(); }
						}
					}, "INIT NEW SOCKET for "+name);
				}
				catch (SocketTimeoutException x) { /** IGNORE **/ }
				catch (Exception x) { x.printStackTrace(); }

				System.out.println("Service "+name+" stopped.");
			}
		}, "SERVICE "+name+" listening on "+SS.getLocalSocketAddress());
	}

	public void listen(final Socket s, final Parser p) throws Exception
	{
		CONTAINER.getDefault().addJob(new Runnable() 
		{
			public void run() 
			{
				while (RUNNING && s.isConnecting() && ! s.isClosed()) try { Thread.sleep(500); } catch (Exception x) { x.printStackTrace(); }

				while (RUNNING && s.isConnected() && !s.isClosed()) try
				{
					Request jo2 = (Request)p.parse();
					Object cmd = jo2.getCommand();
					execute(cmd, jo2, p);
				}
				catch (ReleaseSocketException x)
				{
					break;
				}
				catch (SocketClosedException | SocketException x)
				{
					try { p.close(); } catch (Exception xx) {}
					try { s.close(); } catch (Exception xx) {}
					break;
				}
				catch (Exception x)
				{
					try { p.error(x); } catch (Exception xx) { xx.printStackTrace(); }
					try { p.close(); } catch (Exception xx) { xx.printStackTrace(); }
					try { s.close(); } catch (Exception xx) { xx.printStackTrace(); }
					break;
				}

//				System.out.println("SOCKET "+NAME+" "+s+" listen loop END");
			}
		}, "SOCKET "+NAME+" "+s+" listen loop BEGIN");
	}

	public void on(Object cmd, Callback callback) 
	{
		callbacks.put(cmd, callback);
	}
	
	protected void execute(final Object cmd, final Request data, final Parser parser) throws Exception
	{
/*		
		CONTAINER.getDefault().addJob(new Runnable() 
		{
			@Override
			public void run() 
			{
				try
				{
*/
					Callback cb = callbacks.get(cmd);
					if (cb == null) throw new Exception("No such command: "+cmd);
					parser.execute(data, cb);
/*
				}
				catch (Exception x)
				{
					parser.error(x);
				}
			}
		}, "Executing P2P command");
*/
	}

	public void close() throws IOException
	{
		RUNNING = false;
		SS.close();
	}

	public SocketAddress getLocalSocketAddress() 
	{
		return SS.getLocalSocketAddress();
	}

	public ServerSocket getServerSocket(){
		return SS;
	}
}

File: runtime/botmanager/src/newbound/net/service/service.py
import traceback
import sys
from newbound.robot.botutil import BotUtil
from .surrendersocketexception import SurrenderSocketException

class Service(BotUtil):

    def __init__(self, serversocket, name, parser, container):
        self.serversocket = serversocket
        self.name = name
        self.parser = parser
        self.container = container
        self.running = False
        self.callbacks = {}
        container.addNumThreads(1)
        container.addJob(self.serviceloop, "Service "+self.name+" listening on port "+str(serversocket.getPort()))
        print("Service "+self.name+" started on port "+str(serversocket.getPort()))

    def serviceloop(self):
        self.running = True
        print("Service "+self.name+" starting up...")
        while self.running:
            try:
                self.accept()
            except Exception as e:
                print("Unexpected error listening for "+self.name+" connections "+str(e))
                traceback.print_exc(file=sys.stdout)

    def accept(self):
        s = self.serversocket.accept()
        p = self.parser()
        if p.init(self, s):
            self.listen(s, p)

    def socketloop(self, s, p):
        while (self.running and s.isConnected()):
            try:
                r = p.parse()
                if r == None:
                    s.close()
                else:
                    try:
                        cmd = r.getCommand()
                        self.execute(cmd, r, p)
                    except SurrenderSocketException:
                        break
                    except Exception as e:
                        print("Unexpected error parsing " + self.name + " command " + str(e))
                        traceback.print_exc(file=sys.stdout)
            except Exception as e:
                print("Unexpected error parsing " + self.name + " command " + str(e))
                traceback.print_exc(file=sys.stdout)
                s.close()

    def listen(self, s, p):
        def cb():
            self.socketloop(s, p)
        self.container.addJob(cb, "Socket "+self.name)

    def on(self, cmd, callback):
        self.callbacks[cmd] = callback

    def execute(self, cmd, request, parser):
        if cmd in self.callbacks:
            parser.execute(request, self.callbacks[cmd])
        else:
            print("Ignoring unknown command "+str(cmd)+" / "+str(request.getData()))
        #parser.send(parser.error(cmd, request, 404, "File not found"))

    def close(self):
        self.running = False
        self.serversocket.close()

    def getAddress(self):
        return self.serversocket.getAddress()

    def getPort(self):
        return self.serversocket.getPort()
		